File: startrek-java/StarTrek/src/main/java/chat/dim/socket/BaseChannel.java
/* license: https://mit-license.org
 *
 *  Star Trek: Interstellar Transport
 *
 *                                Written in 2021 by Moky <albert.moky@gmail.com>
 *
 * ==============================================================================
 * The MIT License (MIT)
 *
 * Copyright (c) 2021 Albert Moky
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 * ==============================================================================
 */
package chat.dim.socket;

import java.io.IOException;
import java.net.SocketAddress;
import java.net.SocketException;
import java.nio.ByteBuffer;
import java.nio.channels.ByteChannel;
import java.nio.channels.NetworkChannel;
import java.nio.channels.SelectableChannel;

import chat.dim.net.Channel;
import chat.dim.net.SocketHelper;
import chat.dim.type.AddressPairObject;

public abstract class BaseChannel<C extends SelectableChannel>
        extends AddressPairObject
        implements Channel {

    // socket reader/writer
    protected final SocketReader reader;
    protected final SocketWriter writer;

    // inner socket
    private C impl;
    private int flag;  // closed flag
                       //   -1: initialized
                       //    0: false
                       //    1: true

    /**
     *  Create channel
     *
     * @param remote      - remote address
     * @param local       - local address
     */
    protected BaseChannel(SocketAddress remote, SocketAddress local) {
        super(remote, local);
        reader = createReader();
        writer = createWriter();
        impl = null;
        flag = -1;
    }

    @Override
    protected void finalize() throws Throwable {
        // make sure the relative socket is removed
        setSocketChannel(null);
        super.finalize();
    }

    /**
     *  Create socket reader
     */
    protected abstract SocketReader createReader();

    /**
     *  Create socket writer
     */
    protected abstract SocketWriter createWriter();

    //
    //  Socket
    //

    public C getSocketChannel() {
        return impl;
    }

    /**
     *  Set inner socket for this channel
     */
    public void setSocketChannel(C sock) {
        // 1. replace with new socket
        C old = impl;
        if (sock != null) {
            impl = sock;
            flag = 0;
        } else {
            impl = null;
            flag = 1;
        }
        // 2. close old socket
        if (old != null && old != sock) {
            SocketHelper.socketDisconnect(old);
        }
    }

    //
    //  States
    //

    @Override
    public Status getStatus() {
        if (flag < 0) {
            // initializing
            return Status.INIT;
        }
        C sock = impl;
        if (sock == null || !SocketHelper.socketIsOpen(sock)) {
            // closed
            return Status.CLOSED;
        } else if (SocketHelper.socketIsConnected(sock) || SocketHelper.socketIsBound(sock)) {
            // normal
            return Status.ALIVE;
        } else {
            // opened
            return Status.OPEN;
        }
    }

    @Override
    public boolean isOpen() {
        if (flag < 0) {
            // initializing, it is not 'closed'
            return true;
        }
        C sock = impl;
        return sock != null && SocketHelper.socketIsOpen(sock);
    }

    @Override
    public boolean isBound() {
        C sock = impl;
        return sock != null && SocketHelper.socketIsBound(sock);
    }

    @Override
    public boolean isConnected() {
        C sock = impl;
        return sock != null && SocketHelper.socketIsConnected(sock);
    }

    @Override
    public boolean isAlive() {
        return isOpen() && (isConnected() || isBound());
    }

    @Override
    public boolean isAvailable() {
        C sock = impl;
        if (sock != null && SocketHelper.socketIsOpen(sock)) {
            if (SocketHelper.socketIsConnected(sock) || SocketHelper.socketIsBound(sock)) {
                // alive, check reading buffer
                return checkAvailable(sock);
            }
        }
        return false;
    }

    protected boolean checkAvailable(C sock) {
        return SocketHelper.socketIsAvailable(sock);
    }

    @Override
    public boolean isVacant() {
        C sock = impl;
        if (sock != null && SocketHelper.socketIsOpen(sock)) {
            if (SocketHelper.socketIsConnected(sock) || SocketHelper.socketIsBound(sock)) {
                // alive, check writing buffer
                return checkVacant(sock);
            }
        }
        return false;
    }

    protected boolean checkVacant(C sock) {
        return SocketHelper.socketIsVacant(sock);
    }

    @Override
    public boolean isBlocking() {
        C sock = impl;
        return sock != null && SocketHelper.socketIsBlocking(sock);
    }

    @Override
    public SelectableChannel configureBlocking(boolean block) throws IOException {
        C sock = getSocketChannel();
        if (sock != null) {
            sock.configureBlocking(block);
        }
        return sock;
    }

    @Override
    public String toString() {
        String cname = getClass().getName();
        return "<" + cname + " remote=\"" + getRemoteAddress() + "\" local=\"" + getLocalAddress() + "\"" +
                " closed=" + (!isOpen()) + " bound=" + isBound() + " connected=" + isConnected() + ">\n\t"
                + impl + "\n</" + cname + ">";
    }

    protected boolean bind(C sock, SocketAddress local) {
        if (sock instanceof NetworkChannel) {
            return SocketHelper.socketBind((NetworkChannel) sock, local);
        }
        assert false : "socket error: " + sock;
        return false;
    }

    protected boolean connect(C sock, SocketAddress remote) {
        if (sock instanceof NetworkChannel) {
            return SocketHelper.socketConnect((NetworkChannel) sock, remote);
        }
        assert false : "socket error: " + sock;
        return false;
    }

    protected boolean disconnect(C sock) {
        return SocketHelper.socketDisconnect(sock);
    }

    @Override
    public NetworkChannel bind(SocketAddress local) throws IOException {
        if (local == null) {
            local = localAddress;
            assert local != null : "local address not set";
        }
        C sock = getSocketChannel();
        boolean ok = sock != null && bind(sock, local);
        if (!ok) {
            throw new SocketException("failed to bind socket: " + local);
        }
        localAddress = local;
        return (NetworkChannel) sock;
    }

    @Override
    public NetworkChannel connect(SocketAddress remote) throws IOException {
        if (remote == null) {
            remote = remoteAddress;
            assert remote != null : "remote address not set";
        }
        C sock = getSocketChannel();
        boolean ok = sock != null && connect(sock, remote);
        if (!ok) {
            throw new SocketException("failed to connect socket: " + remote);
        }
        remoteAddress = remote;
        return (NetworkChannel) sock;
    }

    @Override
    public ByteChannel disconnect() throws IOException {
        C sock = impl;
        boolean ok = sock == null || disconnect(sock);
        if (!ok) {
            throw new SocketException("failed to disconnect socket: " + sock);
        }
        // remoteAddress = null;
        return sock instanceof ByteChannel ? (ByteChannel) sock : null;
    }

    @Override
    public void close() throws IOException {
        setSocketChannel(null);
    }

    //
    //  Input/Output
    //

    @Override
    public int read(ByteBuffer dst) throws IOException {
        try {
            return reader.read(dst);
        } catch (IOException e) {
            close();
            throw e;
        }
    }

    @Override
    public int write(ByteBuffer src) throws IOException {
        try {
            return writer.write(src);
        } catch (IOException e) {
            close();
            throw e;
        }
    }

    @Override
    public SocketAddress receive(ByteBuffer dst) throws IOException {
        try {
            return reader.receive(dst);
        } catch (IOException e) {
            close();
            throw e;
        }
    }

    @Override
    public int send(ByteBuffer src, SocketAddress target) throws IOException {
        try {
            return writer.send(src, target);
        } catch (IOException e) {
            close();
            throw e;
        }
    }

}

File: startrek-py/startrek/socket/base_channel.py
# -*- coding: utf-8 -*-
#
#   Star Trek: Interstellar Transport
#
#                                Written in 2021 by Moky <albert.moky@gmail.com>
#
# ==============================================================================
# MIT License
#
# Copyright (c) 2021 Albert Moky
#
# Permission is hereby granted, free of charge, to any person obtaining a copy
# of this software and associated documentation files (the "Software"), to deal
# in the Software without restriction, including without limitation the rights
# to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
# copies of the Software, and to permit persons to whom the Software is
# furnished to do so, subject to the following conditions:
#
# The above copyright notice and this permission notice shall be included in all
# copies or substantial portions of the Software.
#
# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
# IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
# FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
# AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
# LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
# OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
# SOFTWARE.
# ==============================================================================

import socket
from abc import ABC, abstractmethod
from typing import Optional, Tuple

from ..types import SocketAddress, AddressPairObject

from ..net import Channel, ChannelStatus

from .helpers import SocketReader, SocketWriter
from .helpers import SocketHelper


class BaseChannel(AddressPairObject, Channel, ABC):

    def __init__(self, remote: Optional[SocketAddress], local: Optional[SocketAddress]):
        super().__init__(remote=remote, local=local)
        # inner socket
        self.__sock: Optional[socket.socket] = None
        self.__closed = None
        # create socket helper
        self.__helper = self._create_helper()
        # create socket reader/writer
        self.__reader = self._create_reader()
        self.__writer = self._create_writer()

    #
    #   Socket Channel Controllers
    #

    @abstractmethod
    def _create_reader(self) -> SocketReader:
        """ create socket reader """
        raise NotImplemented

    @abstractmethod
    def _create_writer(self) -> SocketWriter:
        """ create socket writer """
        raise NotImplemented

    @property  # protected
    def reader(self) -> SocketReader:
        return self.__reader

    @property  # protected
    def writer(self) -> SocketWriter:
        return self.__writer

    @abstractmethod
    def _create_helper(self) -> SocketHelper:
        raise NotImplemented

    @property  # protected
    def socket_helper(self) -> SocketHelper:
        return self.__helper

    #
    #   socket
    #

    @property
    def sock(self) -> Optional[socket.socket]:
        """ inner socket """
        return self.__sock

    async def set_socket(self, sock: Optional[socket.socket]):
        """ set inner socket for this channel """
        # 1. replace with new socket
        old = self.__sock
        if sock is not None:
            self.__sock = sock
            self.__closed = False
        else:
            self.__sock = None
            self.__closed = True
        # 2. close old socket
        if old is not None and old is not sock:
            helper = self.socket_helper
            await helper.disconnect(sock=old)

    #
    #   States
    #

    @property  # Override
    def status(self) -> ChannelStatus:
        if self.__closed is None:
            # initializing
            return ChannelStatus.INIT
        sock = self.sock
        helper = self.socket_helper
        if sock is None or helper.is_closed(sock=sock):
            # closed
            return ChannelStatus.CLOSED
        elif helper.is_connected(sock=sock) or helper.is_bound(sock=sock):
            # normal
            return ChannelStatus.ALIVE
        else:
            # opened
            return ChannelStatus.OPEN

    @property  # Override
    def closed(self) -> bool:
        if self.__closed is None:
            # initializing
            return False
        sock = self.sock
        helper = self.socket_helper
        return sock is None or helper.is_closed(sock=sock)

    @property  # Override
    def bound(self) -> bool:
        sock = self.sock
        helper = self.socket_helper
        return sock is not None and helper.is_bound(sock=sock)

    @property  # Override
    def connected(self) -> bool:
        sock = self.sock
        helper = self.socket_helper
        return sock is not None and helper.is_connected(sock=sock)

    @property  # Override
    def alive(self) -> bool:
        return (not self.closed) and (self.connected or self.bound)

    @property
    def available(self) -> bool:
        sock = self.sock
        helper = self.socket_helper
        if sock is None or helper.is_closed(sock=sock):
            return False
        elif helper.is_bound(sock=sock) or helper.is_connected(sock=sock):
            # alive, check reading buffer
            return helper.is_available(sock=sock)

    @property
    def vacant(self) -> bool:
        sock = self.sock
        helper = self.socket_helper
        if sock is None or helper.is_closed(sock=sock):
            return False
        elif helper.is_bound(sock=sock) or helper.is_connected(sock=sock):
            # alive, check writing buffer
            return helper.is_vacant(sock=sock)

    @property  # Override
    def blocking(self) -> bool:
        sock = self.sock
        if sock is None:
            return False
        helper = self.socket_helper
        return helper.is_blocking(sock=sock)

    def __str__(self) -> str:
        mod = self.__module__
        cname = self.__class__.__name__
        return '<%s remote="%s" local="%s" closed=%d bound=%d connected=%d >\n\t%s\n</%s module="%s">'\
               % (cname, self._remote, self._local, self.closed, self.bound, self.connected, self.sock, cname, mod)

    def __repr__(self) -> str:
        mod = self.__module__
        cname = self.__class__.__name__
        return '<%s remote="%s" local="%s" closed=%d bound=%d connected=%d >\n\t%s\n</%s module="%s">'\
               % (cname, self._remote, self._local, self.closed, self.bound, self.connected, self.sock, cname, mod)

    # Override
    def configure_blocking(self, blocking: bool):
        sock = self.sock
        helper = self.socket_helper
        helper.set_blocking(sock=sock, blocking=blocking)
        return sock

    # Override
    def bind(self, address: Optional[SocketAddress] = None,
             host: Optional[str] = '0.0.0.0', port: Optional[int] = 0):
        if address is None:
            if port > 0:
                assert host is not None, 'host should not be empty'
                address = (host, port)
            else:
                address = self._local
                assert address is not None, 'local address not set'
        sock = self.sock
        helper = self.socket_helper
        ok = helper.bind(sock=sock, local=address)
        assert ok, 'failed to bind socket: %s' % str(address)
        self._local = address
        return sock

    # Override
    async def connect(self, address: Optional[SocketAddress] = None,
                      host: Optional[str] = '127.0.0.1', port: Optional[int] = 0) -> socket.socket:
        if address is None:
            if port > 0:
                assert host is not None, 'host should not be empty'
                address = (host, port)
            else:
                address = self._remote
                assert address is not None, 'remote address not set'
        sock = self.sock
        helper = self.socket_helper
        ok = await helper.connect(sock=sock, remote=address)
        assert ok, 'failed to connect socket: %s' % str(address)
        self._remote = address
        return sock

    # Override
    async def disconnect(self) -> Optional[socket.socket]:
        sock = self.__sock
        if sock is not None:
            helper = self.socket_helper
            ok = await helper.disconnect(sock=sock)
            assert ok, 'failed to disconnect socket: %s' % sock
            self.__sock = None
        return sock

    # Override
    async def close(self):
        await self.set_socket(sock=None)

    # Override
    async def read(self, max_len: int) -> Optional[bytes]:
        try:
            return await self.reader.read(max_len=max_len)
        except OSError as error:
            await self.close()
            raise error

    # Override
    async def write(self, data: bytes) -> int:
        try:
            return await self.writer.write(data=data)
        except OSError as error:
            await self.close()
            raise error

    # Override
    async def receive(self, max_len: int) -> Tuple[Optional[bytes], Optional[SocketAddress]]:
        try:
            return await self.reader.receive(max_len=max_len)
        except OSError as error:
            await self.close()
            raise error

    # Override
    async def send(self, data: bytes, target: SocketAddress) -> int:
        try:
            return await self.writer.send(data=data, target=target)
        except OSError as error:
            await self.close()
            raise error
